In [1]:
import pandas as pd
from math import log

<b>pandas</b> buat manipulasi <i>data frame</i><br>
<b>math</b> buat ambil/melakukan operasi logaritma

In [2]:
data = pd.read_csv("heart.csv")
data

Age  Gender  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63       1   1       145   233    1        2      150      0      2.3   
1     67       1   4       160   286    0        2      108      1      1.5   
2     67       1   4       120   229    0        2      129      1      2.6   
3     37       1   3       130   250    0        0      187      0      3.5   
4     41       0   2       130   204    0        2      172      0      1.4   
5     56       1   2       120   236    0        0      178      0      0.8   
6     62       0   4       140   268    0        2      160      0      3.6   
7     57       0   4       120   354    0        0      163      1      0.6   
8     63       1   4       130   254    0        2      147      0      1.4   
9     53       1   4       140   203    1        2      155      1      3.1   
10    57       1   4       140   192    0        0      148      0      0.4   
11    56       0   2       140   294    0        2      153      0      1.3   
12    56       1   3       130   256    1        2      142      1      0.6   
13    44       1   2       120   263    0        0      173      0      0.0   
14    52       1   3       172   199    1        0      162      0      0.5   
15    57       1   3       150   168    0        0      174      0      1.6   
16    48       1   2       110   229    0        0      168      0      1.0   
17    54       1   4       140   239    0        0      160      0      1.2   
18    48       0   3       130   275    0        0      139      0      0.2   
19    49       1   2       130   266    0        0      171      0      0.6   
20    64       1   1       110   211    0        2      144      1      1.8   
21    58       0   1       150   283    1        2      162      0      1.0   
22    58       1   2       120   284    0        2      160      0      1.8   
23    58       1   3       132   224    0        2      173      0      3.2   
24    60       1   4       130   206    0        2      132      1      2.4   
25    50       0   3       120   219    0        0      158      0      1.6   
26    58       0   3       120   340    0        0      172      0      0.0   
27    66       0   1       150   226    0        0      114      0      2.6   
28    43       1   4       150   247    0        0      171      0      1.5   
29    40       1   4       110   167    0        2      114      1      2.0   
..   ...     ...  ..       ...   ...  ...      ...      ...    ...      ...   
273   71       0   4       112   149    0        0      125      0      1.6   
274   59       1   1       134   204    0        0      162      0      0.8   
275   64       1   1       170   227    0        2      155      0      0.6   
276   66       0   3       146   278    0        2      152      0      0.0   
277   39       0   3       138   220    0        0      152      0      0.0   
278   57       1   2       154   232    0        2      164      0      0.0   
279   58       0   4       130   197    0        0      131      0      0.6   
280   57       1   4       110   335    0        0      143      1      3.0   
281   47       1   3       130   253    0        0      179      0      0.0   
282   55       0   4       128   205    0        1      130      1      2.0   
283   35       1   2       122   192    0        0      174      0      0.0   
284   61       1   4       148   203    0        0      161      0      0.0   
285   58       1   4       114   318    0        1      140      0      4.4   
286   58       0   4       170   225    1        2      146      1      2.8   
287   58       1   2       125   220    0        0      144      0      0.4   
288   56       1   2       130   221    0        2      163      0      0.0   
289   56       1   2       120   240    0        0      169      0      0.0   
290   67       1   3       152   212    0        2      150      0      0.8   
291   55       0   2       132   342    0        0     

Nilai pada variabel oldpeak kami ubah menjadi bilangan bulat dengan mengalikannya dengan 10 agar memudahkan proses-proses selanjutnya.

In [3]:
data["oldpeak"]=data["oldpeak"]*10
data = data.astype('int64')
data

Age  Gender  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63       1   1       145   233    1        2      150      0       23   
1     67       1   4       160   286    0        2      108      1       15   
2     67       1   4       120   229    0        2      129      1       26   
3     37       1   3       130   250    0        0      187      0       35   
4     41       0   2       130   204    0        2      172      0       14   
5     56       1   2       120   236    0        0      178      0        8   
6     62       0   4       140   268    0        2      160      0       36   
7     57       0   4       120   354    0        0      163      1        6   
8     63       1   4       130   254    0        2      147      0       14   
9     53       1   4       140   203    1        2      155      1       31   
10    57       1   4       140   192    0        0      148      0        4   
11    56       0   2       140   294    0        2      153      0       13   
12    56       1   3       130   256    1        2      142      1        6   
13    44       1   2       120   263    0        0      173      0        0   
14    52       1   3       172   199    1        0      162      0        5   
15    57       1   3       150   168    0        0      174      0       16   
16    48       1   2       110   229    0        0      168      0       10   
17    54       1   4       140   239    0        0      160      0       12   
18    48       0   3       130   275    0        0      139      0        2   
19    49       1   2       130   266    0        0      171      0        6   
20    64       1   1       110   211    0        2      144      1       18   
21    58       0   1       150   283    1        2      162      0       10   
22    58       1   2       120   284    0        2      160      0       18   
23    58       1   3       132   224    0        2      173      0       32   
24    60       1   4       130   206    0        2      132      1       24   
25    50       0   3       120   219    0        0      158      0       16   
26    58       0   3       120   340    0        0      172      0        0   
27    66       0   1       150   226    0        0      114      0       26   
28    43       1   4       150   247    0        0      171      0       15   
29    40       1   4       110   167    0        2      114      1       20   
..   ...     ...  ..       ...   ...  ...      ...      ...    ...      ...   
273   71       0   4       112   149    0        0      125      0       16   
274   59       1   1       134   204    0        0      162      0        8   
275   64       1   1       170   227    0        2      155      0        6   
276   66       0   3       146   278    0        2      152      0        0   
277   39       0   3       138   220    0        0      152      0        0   
278   57       1   2       154   232    0        2      164      0        0   
279   58       0   4       130   197    0        0      131      0        6   
280   57       1   4       110   335    0        0      143      1       30   
281   47       1   3       130   253    0        0      179      0        0   
282   55       0   4       128   205    0        1      130      1       20   
283   35       1   2       122   192    0        0      174      0        0   
284   61       1   4       148   203    0        0      161      0        0   
285   58       1   4       114   318    0        1      140      0       44   
286   58       0   4       170   225    1        2      146      1       28   
287   58       1   2       125   220    0        0      144      0        4   
288   56       1   2       130   221    0        2      163      0        0   
289   56       1   2       120   240    0        0      169      0        0   
290   67       1   3       152   212    0        2      150      0        8   
291   55       0   2       132   342    0        0     

Inti dari praproses pada <i>notebook</i> ini adalah mengubah data <i>continous</i>/non-kategorikal pada variabel <b>Age</b>, <b>trestbps</b>, <b>chol</b>, <b>thalach</b>, dan <b>oldpeak</b> menjadi data kategorikal. Pertama-tama kita harus menentukan berapa banyak kelas/kategori yang dapat dibentuk melalui <i>dataset</i> tersebut, dengan rumus : <b>1 + 3,3 x log(banyak data atau baris)</b><br>

In [4]:
jum_kls = 1+3.3*log(len(data), 10)
jum_kls

9.188760674057605

Selanjutnya adalah menentukan rentang nilai kelas/kategori pada variabel Age, trestbps, chol, thalach, dan oldpeak.

In [5]:
def renNil(jar):
    if (((max(jar)-min(jar))/jum_kls) - int((max(jar)-min(jar))/jum_kls) > 0.5):
        return int((max(jar)-min(jar))/jum_kls)+1
    else:
        return int((max(jar)-min(jar))/jum_kls)

Rentang nilai tiap kelas/kategori pada kolom Age, trestbps, chol, thalach, dan oldpeak yang didapat adalah sebagai berikut :

In [6]:
print(renNil(data["Age"]),
      renNil(data["trestbps"]),
      renNil(data["chol"]),
      renNil(data["thalach"]),
      renNil(data["oldpeak"]))

5 12 48 14 7


Maksudnya seperti ini, mari kita jadikan kolom Age sebagai contoh. Nilai terkecil dari kolom Age adalah 29(tahun) dan nilai terbesar dari kolom tersebut adalah 77(tahun). Dari 29 sampai 77 tersebut terbentuk 9.188760674057605 kategori(sengaja tidak kami bulatkan) di mana tiap kategori memiliki 5 rentang nilai, dengan rincian sebagai berikut :<br><br>
Umur 29-33 tahun masuk ke dalam kelas <b>0</b><br>
Umur 34-38 tahun masuk ke dalam kelas <b>1</b><br>
Umur 39-43 tahun masuk ke dalam kelas <b>2</b><br>
Umur 44-48 tahun masuk ke dalam kelas <b>3</b><br>
Umur 49-53 tahun masuk ke dalam kelas <b>4</b><br>
Umur 54-58 tahun masuk ke dalam kelas <b>5</b><br>
Umur 59-63 tahun masuk ke dalam kelas <b>6</b><br>
Umur 64-68 tahun masuk ke dalam kelas <b>7</b><br>
Umur 69-73 tahun masuk ke dalam kelas <b>8</b><br>
Umur 74-77 tahun masuk ke dalam kelas <b>9</b><br><br>
Hal ini berlaku juga pada variabel/kolom trestbps, chol, thalach, dan oldpeak. Banyaknya kelas/kategori adalah 9.188760674057605 yang jika dibulatkan menjadi 9. Kami sengaja tidak membulatkannya karena jika dibulatkan maka ini akan berpengaruh terhadap banyaknya jumlah anggota pada kelas/kategori tiap kolom dan pada akhirnya hanya terdapat 9 kelas saja. Padahal sebenarnya ada lebih dari 9 kelas. Hal ini kami lakukan untuk menampung nilai yang tersisa pada 0.188760674057605 sebagai satu kelas tambahan. Sehingga total dari banyaknya kelas keseluruhan adalah 10.

In [7]:
list_var = [data["Age"], data["trestbps"], data["chol"], data["thalach"], data["oldpeak"]]
ren_var = [5,12,48,14,7]
baru = [[],[],[],[],[]]

Proses mengubah nilai :

In [8]:
for n in range(len(list_var)):
    for i in list_var[n]:
        if(i <= min(list_var[n])+ren_var[n]):
            baru[n].append(0)
        elif(i <= min(list_var[n])+(ren_var[n]*2-1) and i >= min(list_var[n])+ren_var[n]):
            baru[n].append(1)
        elif(i <= min(list_var[n])+(ren_var[n]*3-1) and i >= min(list_var[n])+ren_var[n]*2):
            baru[n].append(2)
        elif(i <= min(list_var[n])+(ren_var[n]*4-1) and i >= min(list_var[n])+ren_var[n]*3):
            baru[n].append(3)
        elif(i <= min(list_var[n])+(ren_var[n]*5-1) and i >= min(list_var[n])+ren_var[n]*4):
            baru[n].append(4)
        elif(i <= min(list_var[n])+(ren_var[n]*6-1) and i >= min(list_var[n])+ren_var[n]*5):
            baru[n].append(5)
        elif(i <= min(list_var[n])+(ren_var[n]*7-1) and i >= min(list_var[n])+ren_var[n]*6):
            baru[n].append(6)
        elif(i <= min(list_var[n])+(ren_var[n]*8-1) and i >= min(list_var[n])+ren_var[n]*7):
            baru[n].append(7)
        elif(i <= min(list_var[n])+(ren_var[n]*9-1) and i >= min(list_var[n])+ren_var[n]*8):
            baru[n].append(8)
        elif(i <= min(list_var[n])+(ren_var[n]*10-1) and i >= min(list_var[n])+ren_var[n]*9):
            baru[n].append(9)

Masukkan nilai baru ke dalam variabel non kategorikal

In [9]:
data["Age"]      = baru[0]
data["trestbps"] = baru[1]
data["chol"]     = baru[2]
data["thalach"]  = baru[3]
data["oldpeak"]  = baru[4]

In [10]:
data   # Tampilan hasil akhir

Age  Gender  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0      6       1   1         4     2    1        2        5      0        3   
1      7       1   4         5     3    0        2        2      1        2   
2      7       1   4         2     2    0        2        4      1        3   
3      1       1   3         3     2    0        0        8      0        5   
4      2       0   2         3     1    0        2        7      0        2   
5      5       1   2         2     2    0        0        7      0        1   
6      6       0   4         3     2    0        2        6      0        5   
7      5       0   4         2     4    0        0        6      1        0   
8      6       1   4         3     2    0        2        5      0        2   
9      4       1   4         3     1    1        2        6      1        4   
10     5       1   4         3     1    0        0        5      0        0   
11     5       0   2         3     3    0        2        5      0        1   
12     5       1   3         3     2    1        2        5      1        0   
13     3       1   2         2     2    0        0        7      0        0   
14     4       1   3         6     1    1        0        6      0        0   
15     5       1   3         4     0    0        0        7      0        2   
16     3       1   2         1     2    0        0        6      0        1   
17     5       1   4         3     2    0        0        6      0        1   
18     3       0   3         3     3    0        0        4      0        0   
19     4       1   2         3     2    0        0        7      0        0   
20     7       1   1         1     1    0        2        5      1        2   
21     5       0   1         4     3    1        2        6      0        1   
22     5       1   2         2     3    0        2        6      0        2   
23     5       1   3         3     2    0        2        7      0        4   
24     6       1   4         3     1    0        2        4      1        3   
25     4       0   3         2     1    0        0        6      0        2   
26     5       0   3         2     4    0        0        7      0        0   
27     7       0   1         4     2    0        0        3      0        3   
28     2       1   4         4     2    0        0        7      0        2   
29     2       1   4         1     0    0        2        3      1        2   
..   ...     ...  ..       ...   ...  ...      ...      ...    ...      ...   
273    8       0   4         1     0    0        0        3      0        2   
274    6       1   1         3     1    0        0        6      0        1   
275    7       1   1         6     2    0        2        6      0        0   
276    7       0   3         4     3    0        2        5      0        0   
277    2       0   3         3     1    0        0        5      0        0   
278    5       1   2         5     2    0        2        6      0        0   
279    5       0   4         3     1    0        0        4      0        0   
280    5       1   4         1     4    0        0        5      1        4   
281    3       1   3         3     2    0        0        7      0        0   
282    5       0   4         2     1    0        1        4      1        2   
283    1       1   2         2     1    0        0        7      0        0   
284    6       1   4         4     1    0        0        6      0        0   
285    5       1   4         1     4    0        1        4      0        6   
286    5       0   4         6     2    1        2        5      1        4   
287    5       1   2         2     1    0        0        5      0        0   
288    5       1   2         3     1    0        2        6      0        0   
289    5       1   2         2     2    0        0        7      0        0   
290    7       1   3         4     1    0        2        5      0        1   
291    5       0   2         3     4    0        0     

In [11]:
data.to_csv("bersih", sep=',')